atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/InvalidationRunner.java

```diff
@@ -0,0 +1,198 @@
+                    if (tableStatus == TableStatus.NO_DATA) {
+                        lastAllocated = AtlasDbFactory.NO_OP_FAST_FORWARD_TIMESTAMP;
+                    } else {
+                        lastAllocated = limits.upperLimit().value();
+                    }
+
+                    poisonTable(connection);
+                    return lastAllocated;
+                });
+        switch (result.getStatus()) {
+            case SUCCESSFUL:
+                return result.getResultValue();
+            case UNKNOWN:
+            case FAILED:
+                Throwable error = result.getError();
+                if (error instanceof SQLException) {
+                    throw PalantirSqlException.create((SQLException) error);
+                }
+                throw Throwables.rewrapAndThrowUncheckedException(error);
+            default:
+                throw new IllegalStateException("Unrecognized transaction status " + result.getStatus());
+        }
+    }
+
+    private void poisonTable(Connection connection) throws SQLException {
+        try (Statement statement = connection.createStatement()) {
+            if (ConnectionDbTypes.getDbType(connection).equals(DBType.ORACLE)) {
+                statement.execute(String.format("ALTER TABLE %s RENAME COLUMN last_allocated TO LEGACY_last_allocated",
+                        prefixedTimestampTableName()));
+            } else {
+                statement.execute(String.format("ALTER TABLE %s RENAME last_allocated TO LEGACY_last_allocated",
+                        prefixedTimestampTableName()));
+            }
+        }
+    }
+
+    private Limits getLimits(Connection connection) throws SQLException {
+        DatabaseMetaData metaData = connection.getMetaData();
+        ResultSet res = metaData.getTables(null, null, prefixedTimestampTableName(), null);
+
+        Preconditions.checkState(res.next(), "We are in the process of invalidating the "
+                + "InDbTimestampBoundStore but the data table does not exist. "
+                + "We should never reach here. Please contact support.");
+
+        return ImmutableLimits.builder()
+                .upperLimit(getColumnStatus(LAST_ALLOCATED, connection))
+                .legacyUpperLimit(getColumnStatus(LEGACY_LAST_ALLOCATED, connection))
+                .build();
+    }
+
+    private ColumnStatus getColumnStatus(String colName, Connection connection) throws SQLException {
+        if (hasColumn(connection, colName)) {
+            String sql = String.format("SELECT %s FROM %s FOR UPDATE", colName, prefixedTimestampTableName());
+            QueryRunner run = new QueryRunner();
+            return run.query(connection, sql, rs -> {
+                if (rs.next()) {
+                    return ColumnStatus.columnStatusWithValue(rs.getLong(colName));
+                }
+                return ColumnStatus.columnStatusWithoutValue();
+            });
+
+        } else {
+            return ColumnStatus.voidColumnStatus();
```

Comment:
nit: the if above returns, so else here isn't needed.

Reply:
in fact, it's probably nicer to have if(!hasColumn(...)) return ... and then the rest doesn't need to be in an if-block